### PR TITLE
ROX-35509: Create and Populate Vuln Origin

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Adds a new `origin` field to Stackrox vulnerability types. This field will be leveraged in future PRs for showing and filtering CVEs in reports and the UI.

The field is populated by translating the Scanner V4 / Claircore `updater` name. 

The main functional changes are in `pkg/scanners/scannerv4/convert.go`, the remaining changes are passing the new field around, generated code, and tests.

## User-facing documentation

- [x] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [x] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [x] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [x] modified existing tests

### How I validated my change

CI, unit tests, and manual tests as follows:

Forced Scan (bypasses Central DB store/retrieval):

```sh
$ export IMAGE=quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:89db126b1ec3feaef7cddabc3af861860ccc27eed4c86b6c700b90f98657f73c

$ rctl image scan --image=$IMAGE --force 2>/dev/null | jq -r '.scan.components[]?.vulns[]?.origin' | sort | uniq -c
  30 VULN_ORIGIN_OSV
  46 VULN_ORIGIN_RED_HAT
```

Cached Scan (leverages Central DB store/retrieval):

```sh
$ rctl image scan --image=$IMAGE 2>/dev/null | jq -r '.scan.components[]?.vulns[]?.origin' | sort | uniq -c
  30 VULN_ORIGIN_OSV
  46 VULN_ORIGIN_RED_HAT
```

Direct REST API retrieval:

```sh
$ curl -ksS -H "Authorization: Bearer $ROX_API_TOKEN" https://$ROX_ENDPOINT/v1/images/sha256:89db126b1ec3feaef7cddabc3af861860ccc27eed4c86b6c700b90f98657f73c | jq -r '.scan.components[]?.vulns[]?.origin' | sort | uniq -c
  30 VULN_ORIGIN_OSV
  46 VULN_ORIGIN_RED_HAT
```

### CODE REVIEW

File: pkg/scanners/scannerv4/convert.go [hunk 1]
```diff
@@ -776,4 +777,33 @@ func mergeScoringFields(dst, src *storage.EmbeddedVulnerability) {
 	dst.Epss = src.GetEpss()
 	dst.Exploit = src.GetExploit()
 	dst.CisaKev = src.GetExploit() != nil
+	dst.Origin = src.GetOrigin()
```
Comment (jvdm) [inline pkg/scanners/scannerv4/convert.go:876]: How should Origin behave when dedup merges the same CVE from different updaters?

`mergeFixFields` and `mergeScoringFields` can select different source records, but Origin is always set by the highest score. We can therefore return a Originmismatching the fix data. Now, the new Origin field represents the vulnerability feed, not CVSS origin.

File: pkg/scanners/scannerv4/convert_test.go
Comment (jvdm) [inline pkg/scanners/scannerv4/convert_test.go:1]: From the design:

> We will also add an additional test case to catch when a new updater is observed that maps to “Other”. Then, when we observe this test case failing we would patch the origin mapping to map that new updater.

This test only tests hand-picked strings and actually asserts that an unknown updater maps successfully to OTHER. That does not detect introduction of a new real updater.

Comment (dcaravel) [reply pkg/scanners/scannerv4/convert_test.go:1]: ROX-36304 captures this work (same parent feature)

File: pkg/scanners/scannerv4/convert.go (hunk 1 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go:876]: In the current claircore implementation, for a given package, the updater will be the same for all vulns - so the scenario described will not happen. (Also tested this by pulling ~130 vuln reports for images from an OCP cluster and observed zero 'mixed updater' scenarios)

Granted this is not an enforced constraint and may change in the future. And I would expect this to change (ie: when/if we redefine what a 'vulnerability' is, such as a wrapper that references multiple perspectives, at which point deduping is no longer necessary) which will be a foundational change for all of StackRox.

In the event that does happen in the future (different updaters for the same package)- current state the Origin will be aligned with the CVSS and Severity (and other scoring fields)  - your right there chance that the fixed version may come from a different origin from where the CVSS, severity, etc. came from when the different perspectives do not agree on the fixed version. I see this a low impact edge case.

Would you like to see this handled differently?

Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go:876]: > In the current claircore implementation, for a given package, the updater will be the same for all vulns
> [...]
> Would you like to see this handled differently?

I would suggest we flag/capture/comment these assumptions on non-contractual behavior. I feel we're taking the risk of making dependencies on non-contractual claicore interfaces/behaviour (another example is the updater names) in the name of speed, so at least commenting can help.

Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go:876]: > In the current claircore implementation, for a given package, the updater will be the same for all vulns

If that's true, then deduping the origin is irrelevant, right?

Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go:876]: > If that's true, then deduping the origin is irrelevant, right?

Yes, it is meant to be defensive in the event that does change, the output would be deterministic/explainable vs. "the first item in an arbitrary order".

I added comments with explanations. Let me know what you think.

**Review by jvdm — APPROVED** [review]

File: pkg/scanners/scannerv4/convert.go [hunk 2]
```diff
@@ -867,4 +873,37 @@ func mergeScoringFields(dst, src *storage.EmbeddedVulnerability) {
...
+}
+
+var updaterOrigins = []struct {
+	prefix string
+	origin storage.VulnOrigin
+}{
+	{"alpine-", storage.VulnOrigin_VULN_ORIGIN_ALPINE},
```
Comment (alkmim) [inline pkg/scanners/scannerv4/convert.go:883]: Is there a reason to have this hardcoded?

It would be better if we could add another updater without having to edit code.

Comment (dcaravel) [reply pkg/scanners/scannerv4/convert.go:883]: That would be nice. If I'm missing something obvious please let me know.

Not so sure it can be avoided without significantly increasing scope to make this 'data driven' at every touchpoint (UI, reports, API, etc.). The enum value will be translated to 'friendly' names in UI and reports, and we cannot rely on ClairCore internal types directly throughout ACS (as a general best practice but also we support multiple scanners) - which is why we (for example) have similar translations for other fields (ie: severities) to ACS domain types.  And those are later translated to friendly names where applicable. 

To some extent prefixes allow for new matchers to be added without requiring changes (ie: `osv/crates.io` for RUST support), however other changes would still be needed (ie: adding a new `SourceType` for RUST)

Comment (jvdm) [reply pkg/scanners/scannerv4/convert.go:883]: @alkmim:

> Is there a reason to have this hardcoded?

This is hardcoded because in Claircore, today, there isn't a well-defined list of vulnerability feeds (or "sources") that the library supports, and there isn't a well-defined association of that info with the vulnerability objects in the vulnerability report.

The closest we have is the list of updaters. These are the entities in the Claircore that are responsible for reaching to the vulnerability data feeds (one per feed), pulling the data into the Claircore database. It happens the updaters have names, and the names are exposed in the Claircore output, but their format, list, and contract are not well-defined (they might change, drift, have different prefixes, change from Stackrox assumptions, etc).

David's feature is opportunistically using the updater names to map to the concept of vulnerability origin in Stackrox/ACS. He captured a follow-up  ROX-36340.

**Review by alkmim — APPROVED** [review]
